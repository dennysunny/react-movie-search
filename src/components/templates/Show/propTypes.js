import { object, bool } from 'prop-types';

export const contextTypes = {
  router: object.isRequired,
};

export const propTypes = {
  config: object,
  show: object,
  loading: bool,
};

export const defaultProps = {
};
