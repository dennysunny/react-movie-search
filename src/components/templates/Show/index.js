/* eslint-disable camelcase */
import React, { Component } from 'react';
import * as Vibrant from 'node-vibrant';

import { propTypes, contextTypes } from './propTypes';
import Spinner from '../../atoms/Spinner';
import PageTransition from '../../atoms/PageTransition/index';
import ShowProfile from '../../organisms/ShowProfile';

export default class Show extends Component {
  static propTypes = propTypes;

  static contextTypes = contextTypes;

  state = {
    dcolor: [0, 0, 0],
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.loading) {
      this.getPalette(nextProps);
    }
  }

  getPalette = async (nextProps) => {
    const { show: { poster_path }, config: { images } } = nextProps;
    const path = images.base_url + images.poster_sizes[3] + poster_path;
    const palette = await Vibrant.from(path).getPalette();
    // eslint-disable-next-line no-underscore-dangle
    await this.setState({ dcolor: palette.DarkVibrant._rgb });
  };

  handleShowClick = (showId) => {
    const { router } = this.context;
    const { show: { name } } = this.props;
    const media = name ? 'tv' : 'movie';
    router.history.push(`/${media}/${showId}`);
  };

  handleFullCrewClick = () => {
    const { router } = this.context;
    const { show: { id, name } } = this.props;
    const media = name ? 'tv' : 'movie';
    router.history.push(`/${media}/${id}/crew`);
  };

  render() {
    const { loading, show, config } = this.props;
    const { dcolor } = this.state;

    if (loading) {
      return <Spinner />;
    }

    return (
      <div>
        <PageTransition>
          <ShowProfile
            key={show.id}
            config={config}
            show={show}
            dcolor={dcolor}
            handleShowClick={this.handleShowClick}
            handleFullCrewClick={this.handleFullCrewClick}
          />
        </PageTransition>
      </div>
    );
  }
}
