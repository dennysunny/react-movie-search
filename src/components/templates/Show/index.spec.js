import React from 'react';
import { shallow } from 'enzyme';

import ShowProfile from '../../organisms/ShowProfile';
import Show from '.';

const mockProps = {
  config: {
    images: {
      secure_base_url: 'path/to/image',
      poster_sizes: ['1', '2', '3', '4'],
      backdrop_sizes: ['1', '2', '3', '4'],
    },
  },
  show: {
    id: 1,
    title: 'title',
  },
  loading: false,
  genresLoading: false,
  configLoading: false,
};

const context = { router: { history: ['/movie'] } };

describe('Templates/Show', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<Show {...mockProps} />, { context });

    expect(wrapper).toMatchSnapshot();
  });

  it('should render as expected with new props', () => {
    const wrapper = shallow(<Show {...mockProps} />, { context });
    mockProps.show.id = 141052;
    wrapper.setProps({ ...mockProps });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spinner when loading', () => {
    const wrapper = shallow(<Show {...mockProps} loading />, { context });

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle show click', () => {
    const wrapper = shallow(<Show {...mockProps} />, { context });
    const newMovieId = 2;
    wrapper.find(ShowProfile).props().handleShowClick(newMovieId);

    expect(wrapper.context().router.history.includes(`/movie/${newMovieId}`)).toBe(true);
  });

  it('should handle full crew click', () => {
    const wrapper = shallow(<Show {...mockProps} />, { context });
    wrapper.find(ShowProfile).props().handleFullCrewClick();

    expect(wrapper.context().router.history.includes(`/movie/${mockProps.show.id}/crew`)).toBe(true);
  });
});
