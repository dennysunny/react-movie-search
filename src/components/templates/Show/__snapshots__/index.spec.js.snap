// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Templates/Show should render a spinner when loading 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Show
    config={
      Object {
        "images": Object {
          "backdrop_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "poster_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "secure_base_url": "path/to/image",
        },
      }
    }
    configLoading={false}
    genresLoading={false}
    loading={true}
    show={
      Object {
        "id": 141052,
        "title": "title",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "router": Object {
        "history": Array [
          "/movie",
        ],
      },
    },
  },
}
`;

exports[`Templates/Show should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Show
    config={
      Object {
        "images": Object {
          "backdrop_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "poster_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "secure_base_url": "path/to/image",
        },
      }
    }
    configLoading={false}
    genresLoading={false}
    loading={false}
    show={
      Object {
        "id": 1,
        "title": "title",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <PageTransition>
        <ShowProfile
          config={
            Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            }
          }
          dcolor={
            Array [
              0,
              0,
              0,
            ]
          }
          handleFullCrewClick={[Function]}
          handleShowClick={[Function]}
          show={
            Object {
              "id": 1,
              "title": "title",
            }
          }
        />
      </PageTransition>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <ShowProfile
          config={
            Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            }
          }
          dcolor={
            Array [
              0,
              0,
              0,
            ]
          }
          handleFullCrewClick={[Function]}
          handleShowClick={[Function]}
          show={
            Object {
              "id": 1,
              "title": "title",
            }
          }
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "config": Object {
            "images": Object {
              "backdrop_sizes": Array [
                "1",
                "2",
                "3",
                "4",
              ],
              "poster_sizes": Array [
                "1",
                "2",
                "3",
                "4",
              ],
              "secure_base_url": "path/to/image",
            },
          },
          "dcolor": Array [
            0,
            0,
            0,
          ],
          "handleFullCrewClick": [Function],
          "handleShowClick": [Function],
          "show": Object {
            "id": 1,
            "title": "title",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <PageTransition>
          <ShowProfile
            config={
              Object {
                "images": Object {
                  "backdrop_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "poster_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "secure_base_url": "path/to/image",
                },
              }
            }
            dcolor={
              Array [
                0,
                0,
                0,
              ]
            }
            handleFullCrewClick={[Function]}
            handleShowClick={[Function]}
            show={
              Object {
                "id": 1,
                "title": "title",
              }
            }
          />
        </PageTransition>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <ShowProfile
            config={
              Object {
                "images": Object {
                  "backdrop_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "poster_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "secure_base_url": "path/to/image",
                },
              }
            }
            dcolor={
              Array [
                0,
                0,
                0,
              ]
            }
            handleFullCrewClick={[Function]}
            handleShowClick={[Function]}
            show={
              Object {
                "id": 1,
                "title": "title",
              }
            }
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "config": Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            },
            "dcolor": Array [
              0,
              0,
              0,
            ],
            "handleFullCrewClick": [Function],
            "handleShowClick": [Function],
            "show": Object {
              "id": 1,
              "title": "title",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "router": Object {
        "history": Array [
          "/movie",
        ],
      },
    },
  },
}
`;

exports[`Templates/Show should render as expected with new props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Show
    config={
      Object {
        "images": Object {
          "backdrop_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "poster_sizes": Array [
            "1",
            "2",
            "3",
            "4",
          ],
          "secure_base_url": "path/to/image",
        },
      }
    }
    configLoading={false}
    genresLoading={false}
    loading={false}
    show={
      Object {
        "id": 141052,
        "title": "title",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <PageTransition>
        <ShowProfile
          config={
            Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            }
          }
          dcolor={
            Array [
              0,
              0,
              0,
            ]
          }
          handleFullCrewClick={[Function]}
          handleShowClick={[Function]}
          show={
            Object {
              "id": 141052,
              "title": "title",
            }
          }
        />
      </PageTransition>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <ShowProfile
          config={
            Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            }
          }
          dcolor={
            Array [
              0,
              0,
              0,
            ]
          }
          handleFullCrewClick={[Function]}
          handleShowClick={[Function]}
          show={
            Object {
              "id": 141052,
              "title": "title",
            }
          }
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": "141052",
        "nodeType": "class",
        "props": Object {
          "config": Object {
            "images": Object {
              "backdrop_sizes": Array [
                "1",
                "2",
                "3",
                "4",
              ],
              "poster_sizes": Array [
                "1",
                "2",
                "3",
                "4",
              ],
              "secure_base_url": "path/to/image",
            },
          },
          "dcolor": Array [
            0,
            0,
            0,
          ],
          "handleFullCrewClick": [Function],
          "handleShowClick": [Function],
          "show": Object {
            "id": 141052,
            "title": "title",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <PageTransition>
          <ShowProfile
            config={
              Object {
                "images": Object {
                  "backdrop_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "poster_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "secure_base_url": "path/to/image",
                },
              }
            }
            dcolor={
              Array [
                0,
                0,
                0,
              ]
            }
            handleFullCrewClick={[Function]}
            handleShowClick={[Function]}
            show={
              Object {
                "id": 141052,
                "title": "title",
              }
            }
          />
        </PageTransition>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <ShowProfile
            config={
              Object {
                "images": Object {
                  "backdrop_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "poster_sizes": Array [
                    "1",
                    "2",
                    "3",
                    "4",
                  ],
                  "secure_base_url": "path/to/image",
                },
              }
            }
            dcolor={
              Array [
                0,
                0,
                0,
              ]
            }
            handleFullCrewClick={[Function]}
            handleShowClick={[Function]}
            show={
              Object {
                "id": 141052,
                "title": "title",
              }
            }
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "141052",
          "nodeType": "class",
          "props": Object {
            "config": Object {
              "images": Object {
                "backdrop_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "poster_sizes": Array [
                  "1",
                  "2",
                  "3",
                  "4",
                ],
                "secure_base_url": "path/to/image",
              },
            },
            "dcolor": Array [
              0,
              0,
              0,
            ],
            "handleFullCrewClick": [Function],
            "handleShowClick": [Function],
            "show": Object {
              "id": 141052,
              "title": "title",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "router": Object {
        "history": Array [
          "/movie",
        ],
      },
    },
  },
}
`;
